src/common/tusb_fifo.c

```diff
@@ -442,18 +448,19 @@ static uint16_t _tu_fifo_peek_n(tu_fifo_t* f, void * p_buffer, uint16_t n, uint1
   // Check if we can read something at and after offset - if too less is available we read what remains
   if (cnt < n) n = cnt;
 
-  uint16_t rRel = get_relative_pointer(f, rAbs);
+  uint16_t rd_ptr = idx2ptr(rd_idx, f->depth);
 
   // Peek data
-  _ff_pull_n(f, p_buffer, n, rRel, copy_mode);
+  _ff_pull_n(f, p_buffer, n, rd_ptr, copy_mode);
 
   return n;
 }
 
 // Works on local copies of w and r
-static inline uint16_t _tu_fifo_remaining(tu_fifo_t* f, uint16_t wAbs, uint16_t rAbs)
+static inline uint16_t _tu_fifo_remaining(tu_fifo_t* f, uint16_t wr_idx, uint16_t rd_idx)
 {
-  return f->depth - _tu_fifo_count(f, wAbs, rAbs);
+  uint16_t const count = _tu_fifo_count(f, wr_idx, rd_idx);
+  return (f->depth > count) ? (f->depth - count) : 0;
```

Comment:
Does this prevent the out of bound error to happen? Would it not be better to take care `_tu_fifo_count()` does not return a value bigger than `f->depth`? Then you could avoid the overflow check here and at the same time sanitize `_tu_fifo_count()`! Also `f->depth == count` should be ok, this means the buffer is exactly full!

Reply:
yes, this prevent out of bound to occur. I think we can use the public version of tu_fifo_count() which does limit the count to depth. The internal version _tu_fifo_count() seems to be used to determine an overflowed condition so that we could correct/adjust read index in a fifo_read() function.